src/Core/Addon/Module/ModuleManager.php

```diff
@@ -386,12 +417,38 @@ public function upgrade($name, $version = 'latest', $source = null)
         $this->checkIsInstalled($name);
         $module = $this->moduleRepository->getModule($name);
 
-        // Get new module
-        // 1- From source
-        if ($source != null) {
-            $this->moduleZipManager->storeInModulesFolder($source);
+        if ($source === null) {
+            // Give a chance to the other modules (like marketplace) to retrieve the module source
+            $sourceZip = $this->hookManager->exec(
+                'actionAdminModuleUpgradeRetrieveSource',
+                [
+                    'name' => $name,
+                    'current_version' => $module->get('version'),
+                ],
+                null,
+                true
+            );
+
+            if (is_array($sourceZip)) {
+                $sourceZip = array_values($sourceZip)[0];
```

Comment:
It's ok, but I would do without an extra array and hard coded index 0.
Returns `false` if array is empty, as safety measure.
```suggestion
                $sourceZip = reset($sourceZip);
```